source/util/timer.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <sys/resource.h>
+#include <sys/time.h>
+#include <iomanip>
+#include <iostream>
+
+#include "util/timer.h"
+
+namespace spvutils {
+
+#if defined(SPIRV_TIMER_ENABLED)
+
+namespace {
+inline double TimeDifference(const timeval& before, const timeval& after) {
+  return static_cast<double>(after.tv_sec - before.tv_sec) +
+         static_cast<double>((after.tv_usec - before.tv_usec) / 10000) * .01;
+}
+}  // namespace
+
+void TimerPrintDescription(std::ostream* out) {
+  if (out) {
+    *out << std::setw(30) << "PASS name" << std::setw(12) << "USR time"
+         << std::setw(12) << "WALL time" << std::setw(12) << "SYS time"
+#if defined(SPIRV_MEMORY_MEASUREMENT_ENABLED)
+         << std::setw(12) << "RSS" << std::setw(12) << "Pagefault"
+#endif  // defined(SPIRV_MEMORY_MEASUREMENT_ENABLED)
+         << std::endl;
+  }
+}
+
+void Timer::Start() {
+  if (report_stream_) {
+    if (getrusage(RUSAGE_SELF, &usage_before) == -1) {
+      usage_fail = kGetrusageFail;
+    } else if (gettimeofday(&wall_before, NULL) == -1) {
+      usage_fail = kGettimeofdayFail;
+    }
+  }
+}
+
+void Timer::Stop() {
+  if (report_stream_ && usage_fail == kNotFailed) {
+    if (getrusage(RUSAGE_SELF, &usage_after) == -1) {
+      usage_fail = kGetrusageFail;
+    } else if (gettimeofday(&wall_after, NULL) == -1) {
```

Comment:
Why do we need this fallback?

Reply:
getrusage(), gettimeofday(), times(), and clock_gettime() returns -1 when they fail. This code set the usage status as failure if one of those calls fail. Otherwise, it proceeds with the next call.